File: ai/simple.py
import random

from planetwars.ai import planetwars_ai
from planetwars.datatypes import Order
from planetwars.utils import *

@planetwars_ai("StrongToWeak")
def strong_to_weak(turn, pid, planets, fleets):
    my_planets, their_planets, _ = aggro_partition(pid, planets)
    my_strongest = max(my_planets, key=get_ships)
    their_weakest = min(their_planets, key=get_ships)
    return [Order(my_strongest, their_weakest, my_strongest.ships * 0.75)]

@planetwars_ai("AllToWeak")
def all_to_weak(turn, pid, planets, fleets):
    my_planets, their_planets, _ = aggro_partition(pid, planets)
    destination = min(their_planets, key=get_ships)
    orders = []
    for planet in my_planets:
        orders.append(Order(planet, destination, planet.ships * 0.75))
    return orders

@planetwars_ai("AllToCloseOrWeak")
def all_to_close_or_weak(turn, pid, planets, fleets):
    my_planets, their_planets, neutral_planets = aggro_partition(pid, planets)
    other_planets = their_planets + neutral_planets
    their_weakest = min(their_planets, key=get_ships)
    my_total = sum(map(get_ships, my_planets))
    destination = min(their_planets, key=get_ships)
    orders = []
    for planet in my_planets:
        if random.random() < 0.5:
            def dist_to(other_planet):
                return turn_dist(planet, other_planet)
            closest = min(other_planets, key=dist_to)
            orders.append(Order(planet, closest, planet.ships * 0.75))
        else:
            orders.append(Order(planet, their_weakest, planet.ships * 0.75))
    return orders

@planetwars_ai("Random")
def random_ai(turn, pid, planets, fleets):
    def mine(x):
        return x.owner == pid
    my_planets, other_planets = partition(mine, planets)
    source = random.choice(my_planets)
    destination = random.choice(other_planets)
    return [Order(source, destination, source.ships / 2)]

File: ai/simple.js
const partition = require('lodash/partition');

const { Order } = require('../dataTypes');
const { aggroPartition, getShips, turnDist } = require('../utils');
const { max, min, randomChoice } = require('../base');

class StrongToWeak {
  play(turn, player, planets) {
    let [myPlanets, theirPlanets] = aggroPartition(player, planets);
    let myStrongest = max(myPlanets, getShips);
    let theirWeakest = min(theirPlanets, getShips);
    return [new Order(myStrongest, theirWeakest, myStrongest.ships * 0.69)];
  }
}

class Random {
  play(turn, player, planets) {
    let mine = (thing) => thing.owner === player;
    let [myPlanets, otherPlanets] = partition(planets, mine);
    let source = randomChoice(myPlanets);
    let dest = randomChoice(otherPlanets);
    return [new Order(source, dest, source.ships / 2)];
  }
}



class AllToCloseOrWeak {
  play(turn, player, planets) {
    let currentPlanet;
    let orders = [];
    let [myPlanets, theirPlanets] = aggroPartition(player, planets);
    function distTo(otherPlanet){
      return turnDist(currentPlanet,otherPlanet);
    }
    let theirWeakest = min(theirPlanets, getShips);
    myPlanets.forEach(function(planet) {
      currentPlanet = planet;
      if (Math.random() > 0.5) {
        let closest = min(theirPlanets, distTo);
        orders.push(new Order(planet, closest, planet.ships * 0.75));
      } else {
        orders.push(new Order(planet, theirWeakest, planet.ships * 0.75));
      }
    });

    return orders;
  }
}

module.exports = { Random, StrongToWeak, AllToCloseOrWeak };


